### PR TITLE
provider/google: serialized server groups, load balancers, security groups

### PR DESCRIPTION
**Pull request by nwwebb**
This commit has logic to serialize the server groups, autoscalers, and load balancers of an application. It also serializes all of the security groups. I have tested this with several example applications and Terraform is able to correctly interpret the output and apply the serialized infrastructure to GCE. I still need to serialize the static ip addresses and make it so only the security groups within the relevant scope are serialized. After this, I plan to store the output in a GCS bucket with versioning enabled so that the serialized infrastructure is easily accessible and documented for the user. Next I will work on surfacing this functionality to deck as a pipeline stage. Then I will focus on deserializing the infrastructure.
@lwander  @duftler

### CODE REVIEW

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.groovy [hunk 1]
```diff
@@ -57,4 +57,8 @@ final class AtomicOperations {
   public static final String RUN_JOB = "runJob"
   public static final String DESTROY_JOB = "destroyJob"
   public static final String CLONE_JOB = "cloneJob"
+
+  // Serialize operations
+  public static final String SERIALIZE_APPLICATION = "serializeApplication"
+  public static final String DESERIALIZE_APPLICATION = "deserializeApplication"
```
Comment (duftler) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.groovy]: If there is no `DESERIALIZE_APPLICATION` operation yet, we don't need the constant.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/description/SerializeApplicationDescription/SerializeApplicationDescription.groovy [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.google.deploy.description.SerializeApplicationDescription
+
+import com.netflix.spinnaker.clouddriver.google.deploy.description.AbstractGoogleCredentialsDescription
+
+class SerializeApplicationDescription extends AbstractGoogleCredentialsDescription {
+  //TODO: describe the information needed to serialize an application
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/description/SerializeApplicationDescription/SerializeApplicationDescription.groovy]: Please include your github id in the `TODO`. Also, not sure what 'the information needed' refers to exactly. Maybe include a list of the potential attributes?

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/ModifyGoogleServerGroupInstanceTemplateAtomicOperation.groovy [hunk 3]
```diff
@@ -88,14 +88,14 @@ class ModifyGoogleServerGroupInstanceTemplateAtomicOperation implements AtomicOp
     // Will return null if this is a regional server group.
     def zone = serverGroup.zone
 
-    def instanceGroupManagers = isRegional ? compute.regionInstanceGroupManagers() : compute.instanceGroupManagers()
+    def instanceGroupManagers =
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/ModifyGoogleServerGroupInstanceTemplateAtomicOperation.groovy]: The changes in this class look like unintentional reformatting.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+import com.netflix.spinnaker.clouddriver.google.provider.view.GoogleLoadBalancerProvider
+import com.netflix.spinnaker.clouddriver.google.provider.view.GoogleSecurityGroupProvider
+import com.netflix.spinnaker.clouddriver.model.securitygroups.Rule
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import groovy.json.JsonBuilder
+import org.springframework.beans.factory.annotation.Autowired
+
+class SerializeApplicationAtomicOperation implements AtomicOperation<Void> {
+
+  private static final String BASE_PHASE = "SERIALZIE_APPLICATION"
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: s/SERIALZIE/SERIALIZE

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  private final description
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: `string` like its neighbors?

Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Ah, should be `SerializeApplicationDescription`.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 6]
```diff
@@ -0,0 +1,645 @@
...
+  private final String applicationName
+  private final String accountName
+
+  private int numInstanceGroupManagers = 0
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Don't really need these initialization values. The `int`'s will default to `0` anyway.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 7]
```diff
@@ -0,0 +1,645 @@
...
+    this.applicationName = description.applicationName
+    this.accountName = description.accountName
+  }
+
+
+  /* curl -X POST -H "Content-Type: application/json" -d '[ { "serializeApplication": { "applicationName": "codelab", "credentials": "my-google-account" }} ]' localhost:7002/gce/ops */
+  @Override
+  Void operate(List priorOutputs) {
+
+    //TODO: static ip addresses
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Please add your id to this `TODO`.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 8]
```diff
@@ -0,0 +1,645 @@
...
+      cluster.serverGroups.each { GoogleServerGroup.View serverGroup ->
+        addServerGroupToResourceMap(serverGroup, resourceMap)
+      }
+    }
+
+    googleLoadBalancerProvider.getApplicationLoadBalancers(applicationName).each { GoogleLoadBalancer.View loadBalancer ->
+      addLoadBalancerToResourceMap(loadBalancer, resourceMap)
+    }
+
+    //TODO: get only the appropriate security groups rather than all
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Please add your id to this `TODO`.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 9]
```diff
@@ -0,0 +1,645 @@
...
+  }
+
+  private Void addNetworkInterfacesToInstanceTemplateMap(List<NetworkInterface> networkInterfaces, Map instanceTemplateMap) {
+    instanceTemplateMap.network_interface = []
+
+    networkInterfaces.each { NetworkInterface networkInterface ->
+      def networkInterfaceMap = [:]
+      if (networkInterface.network) {
+        String networkUrl = networkInterface.network
+        //TODO: too hacky for name?
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Please add your id to this `TODO`.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 10]
```diff
@@ -0,0 +1,645 @@
...
+        networkInterfaceMap.network = networkUrl.split("/").last()
+      }
+      if (networkInterface.subnetwork) {
+        networkInterfaceMap.subnetwork = networkInterface.subnetwork
+      }
+
+      //TODO(nwwebb) deal with possible static ip here, empty map means ephimeral will be used
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: s/ephimeral/ephemeral

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 11]
```diff
@@ -0,0 +1,645 @@
...
+    if (serverGroup.imageSummary.image) {
+      Map instanceTemplateMap = serverGroup.imageSummary.image
+      InstanceTemplate instanceTemplate = convertMapToInstanceTemplate(instanceTemplateMap)
+      if (!addInstanceTemplateToResourceMap(instanceTemplate, resourceMap)) {
+        return false
+      }
+      serverGroupMap.instance_template = "\${google_compute_instance_template.${instanceTemplate.name}.self_link}"
+    }
+
+    //TODO: get project from account
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Please add your id to these `TODO`s.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 12]
```diff
@@ -0,0 +1,645 @@
...
+    }
+
+    numFirewalls++
+    resourceMap.google_compute_firewall[securityGroup.name] = firewallMap
+    return true
+
+  }
+
+  /*
+   * Conversion for the instance template map, hopefully will find a better way to do this in the future.
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: `ObjectMapper` is probably a better way to do these kinds of conversions. It is used elsewhere in the codebase.

Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: ObjectMapper fails because it can't seem to deserialize nested classes, and has trouble with constant conversions like float -> bigInt.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 13]
```diff
@@ -0,0 +1,645 @@
...
+      addSecurityGroupToResourceMap(securityGroup,resourceMap)
+    }
+
+    cleanUpResourceMap(resourceMap)
+
+    def json = new JsonBuilder()
+    def applicationMap = [:]
+    applicationMap.resource = resourceMap
+    json(applicationMap)
+    println json.toPrettyString()
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Leave a comment that this is temporary.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 14]
```diff
@@ -0,0 +1,645 @@
...
+    resourceMap.google_compute_forwarding_rule = [:]
+    resourceMap.google_compute_http_health_check = [:]
+    resourceMap.google_compute_autoscaler = [:]
+    resourceMap.google_compute_firewall = [:]
+
+    return null
+  }
+
+  /*
+  Need to do this or terraform complains
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]:
```
/*
 *
 */
```

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 15]
```diff
@@ -0,0 +1,645 @@
...
+   */
+  private Boolean addServerGroupToResourceMap(GoogleServerGroup.View serverGroup, Map resourceMap) {
+
+    def serverGroupMap = [:]
+    if (serverGroup.name) {
+      serverGroupMap.name = serverGroup.name
+      serverGroupMap.base_instance_name = serverGroup.name
+    } else {
+      //required
+      return false
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: If this is required (and will fail without it), it's probably better to throw an exception with a descriptive reason here. It will be surfaced by Gate when the operation fails.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy (hunk 12 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: We are handling nested classes here (https://github.com/spinnaker/clouddriver/blob/master/clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/view/GoogleInstanceProvider.groovy#L107). `GoogleLoadBalancer` has a `GoogleHealthCheck` field.

Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: We work around the the bigint issue here: https://github.com/spinnaker/clouddriver/blob/master/clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/view/GoogleSubnetProvider.groovy#L65-L67

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy (hunk 15 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Yup.
Please make these `void` type methods and throw exceptions when issues are encountered. Doesn't look the returned `boolean`s were being considered anyway.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 16]
```diff
@@ -0,0 +1,645 @@
...
+  SerializeApplicationAtomicOperation(SerializeApplicationDescription description) {
+    this.description = description
+    this.applicationName = description.applicationName
+    this.accountName = description.accountName
+  }
+
+
+  /* curl -X POST -H "Content-Type: application/json" -d '[ { "serializeApplication": { "applicationName": "codelab", "credentials": "my-google-account" }} ]' localhost:7002/gce/ops */
+  @Override
+  Void operate(List priorOutputs) {
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: You should emit status messages as the operation proceeds. This operation has quite a few status messages that can serve as examples: https://github.com/spinnaker/clouddriver/blob/master/clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleLoadBalancerAtomicOperation.groovy

Search that class for `updateStatus`.

When the requestor of the operation queries for task status, they will be returned the accumulating status messages (and any exceptions).

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy (hunk 12 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: We talked about this off-line and it seems like ObjectMapper chokes on one of the classes.

Wondering if it really adds much to convert the maps into their original model classes anyway. Maybe just operate directly on the maps? Still working through the PR, so maybe the motivation will become apparent as I read.

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.groovy (hunk 1 above)
Comment (duftler) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.groovy]: Would it make sense to support operating on Spinnaker 'projects' as well? They allow the user to collect Applications/Clusters/Pipelines, and that may make for a nice grouping mechanism to use when determining scope.

We don't need to answer this question necessarily while working on this PR, just wanted to mention it for now.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/model/GoogleLoadBalancer.groovy [hunk 17]
```diff
@@ -33,6 +33,7 @@ class GoogleLoadBalancer {
   String ipAddress
   String ipProtocol
   String portRange
+  String targetPoolLink
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/model/GoogleLoadBalancer.groovy]: Maybe just `targetPool` or `targetPoolUrl`?

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/model/GoogleHealthCheck.groovy [hunk 18]
```diff
@@ -45,6 +45,12 @@ class GoogleHealthCheck {
     int unhealthyThreshold = GoogleHealthCheck.this.unhealthyThreshold
     int healthyThreshold = GoogleHealthCheck.this.healthyThreshold
 
+    //QUESTION: can I add things to view, is this how to do it?
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/model/GoogleHealthCheck.groovy]: The various View classes are used by the Providers to return representations of the underlying resources, without having to return _everything_. Basically a normalized view of the model.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/model/GoogleServerGroup.groovy [hunk 19]
```diff
@@ -45,6 +45,7 @@ class GoogleServerGroup {
   String networkName
   Set<String> instanceTemplateTags = []
   String selfLink
+  Set<String> targetPools = []
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/model/GoogleServerGroup.groovy]: Why is this needed?

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 20]
```diff
@@ -0,0 +1,645 @@
...
+
+    //TODO: get project from account
+    //TODO: named_port, not sure what this is
+    //TODO: update_strategy, automatically restart is default should this be used always (nothing specified in serverGroup object)?
+    if (serverGroup.instanceCounts?.total) {
+      serverGroupMap.target_size = serverGroup.instanceCounts.total
+    }
+
+    if (serverGroup.loadBalancers && !serverGroup.loadBalancers.isEmpty()) {
+      serverGroupMap.target_pools = serverGroup.targetPools
```
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: I think this is the same set of url's you can find by collecting all the google load balancers' `targetPoolUrl`s.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy (hunk 6 above)
Comment (duftler) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: After reading further, it looks like these counters just track the size of the various sub-maps. You can retrieve that size directly from each map, so these counters probably aren't needed.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 21]
```diff
@@ -0,0 +1,671 @@
...
+
+    if (serverGroup.imageSummary.image) {
+      def instanceTemplateMap = serverGroup.imageSummary.image
+      InstanceTemplate instanceTemplate = convertMapToInstanceTemplate(instanceTemplateMap)
+      addInstanceTemplateToResourceMap(instanceTemplate, resourceMap)
+      serverGroupMap.instance_template = "\${google_compute_instance_template.${instanceTemplate.name}.self_link}"
+    }
+
+    serverGroupMap.project = this.project
+    //TODO(nwwebb) update_strategy, automatically restart is default should this be used always?
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Ideally keep it identical to whatever is created using Spinnaker.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/exception/GoogleResourceImproperFormat.groovy [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.google.deploy.exception
+
+import groovy.transform.InheritConstructors
+
+@InheritConstructors
+
+class GoogleResourceImproperFormat extends GoogleOperationException {}
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/exception/GoogleResourceImproperFormat.groovy]: Is this thrown when something is in an illegal/impossible state, in a state not known/understood by Spinnaker, or both?

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 23]
```diff
@@ -0,0 +1,671 @@
...
+      def networkInterfaceMap = [:]
+      if (networkInterface.network) {
+        String networkUrl = networkInterface.network
+        networkInterfaceMap.network = networkUrl.split("/").last()
+      }
+      if (networkInterface.subnetwork) {
+        networkInterfaceMap.subnetwork = networkInterface.subnetwork
+      }
+
+      //TODO(nwwebb) deal with possible static ip here, empty map means ephemeral will be used
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: @duftler do we allow instances to be created with statically allocated ips?

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 24]
```diff
@@ -0,0 +1,671 @@
...
+    }
+    if (autoscalingPolicy.minNumReplicas) {
+      autoscalerMap.autoscaling_policy.min_replicas = autoscalingPolicy.minNumReplicas
+    } else {
+      throw new GoogleResourceImproperFormat("Required minimum number of replicas not found for autoscaler: ${autoscalerMap.name}")
+    }
+    if (autoscalingPolicy.coolDownPeriodSec) {
+      autoscalerMap.autoscaling_policy.cooldown_period = autoscalingPolicy.coolDownPeriodSec
+    }
+    //TODO(nwwebb) terraform only lets you select one policy even though gce lets you select multiple
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Good catch. Feel free to file an issue here: https://github.com/hashicorp/terraform/issues

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 25]
```diff
@@ -0,0 +1,671 @@
...
+          metricMap.name = metric.metric
+        } else {
+          return
+        }
+        if (metric.utilizationTarget) {
+          metricMap.target = metric.utilizationTarget
+        } else {
+          return
+        }
+        //TODO(nwwebb) ask about this, gce doesn't match terraform types
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Another good time to file an issue: https://github.com/hashicorp/terraform/issues

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy [hunk 26]
```diff
@@ -0,0 +1,671 @@
...
+      throw new GoogleResourceImproperFormat("Required health check name not found")
+    }
+    if (healthCheck.interval) {
+      healthCheckMap.check_interval_sec = healthCheck.interval
+    }
+
+    if (healthCheck.healthyThreshold) {
+      healthCheckMap.healthy_threshold = healthCheck.healthyThreshold
+    }
+    //TODO(nwwebb) host
```
Comment (lwander) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy]: Can you elaborate on this TODO?

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/exception/GoogleResourceImproperFormat.groovy (hunk 22 above)
Comment (nwwebb) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/exception/GoogleResourceImproperFormat.groovy]: illlegal/ impossible state, I'll change the name of the exception

**Comment by nwwebb** [comment]
@lwander

**Comment by nwwebb** [comment]
@lwander let me know how this looks

**Comment by lwander** [comment]
LGTM. Let's rebase & merge.

**Comment by nwwebb** [comment]
@lwander added tests for the atomic operation and converter

**Comment by lwander** [comment]
Really nice, good work. I would add a test that tries to serialize misconfigured/illegal infrastructure and correctly throws an exception. You can test this with

```groovy
def e = thrown(GoogleResourceIllegalState)
e.message == 'some message saying what\'s wrong'
```

**Comment by nwwebb** [comment]
@lwander added exception test

**Comment by lwander** [comment]
Nice, LGTM
